services/multitenant/channelCounter.go

```diff
@@ -0,0 +1,52 @@
+package multitenant
+
+type Counter struct {
+	ch                    chan func()
```

Comment:
is the purpose of this channel to help with concurrent map access? If so, should we use a `sync.RWMutex` instead as proposed [here](https://go.dev/blog/maps#concurrency)?

Reply:
it is for concurrent map access, as well as to maintain the order of that access. I read a bit about "mutex fairness" and it didn't seem like ordering was maintained.
Whereas with this, I'm fairly certain the order of access is maintained.

Reply:
what is the concern?

Reply:
one concern is that too many goroutines racing for the mutex around (de)incrementing the count. This is slowing down the processes on those goroutines(processor storing to rt, brt queues).

I suspect this is somehow leading to subtract(terminal counts) before we even add them to pending count.
Hence an attempt to use channels. I'll give the atomic operation(prometheus) you suggested and look for improvements.

Reply:
Incrementing & decrementing an in-memory counter shouldn't be considered to be a cause of delays regardless the number of competing writers. I believe that it is more important here to have readable code and consistent behaviour, i.e. ensure that the counter is safe for concurrent access.

Reply:
The counter is safe for concurrent access right?
while the +/- operation itself isn't delaying, wouldn't taking locks be a cause for that..?

Reply:
> The counter is safe for concurrent access right?

which counter?

Reply:
the map of counts(`workspaceDestCountMap`) in `Counter` struct.

Reply:
yes, looks to be safe. However it seems to be a complex pattern to reuse.